File: rust/src/generic/int_mod_p.rs
use std::fmt;
use crate::generic::i_field::IField;
use crate::generic::i_ordered::IOrdered;
use crate::generic::i_math::IMath;
use crate::generic::i_primitive_roots::IPrimitiveRoots;
use std::hash::Hash;
use std::cmp::Eq;
#[derive(Debug)]
pub struct IntModP {
    pub i: u64,
}

pub static mut MODULUS: u64 = 7;

fn get_modulus() -> u64 {
    unsafe { MODULUS }
}

pub fn set_modulus(p: u64) {
    unsafe {
        MODULUS = p;
    }
}

fn mod_inverse(a: u64, p: u64) -> u64 {
    let (mut t, mut new_t) = (0 as i64, 1 as i64);
    let (mut r, mut new_r) = (p as i64, a.rem_euclid(p) as i64);
    while new_r != 0 {
        let quotient = r / new_r;
        let temp_t = t - quotient * new_t;
        t = new_t;
        new_t = temp_t;
        let temp_r = r - quotient * new_r;
        r = new_r;
        new_r = temp_r;
    }
    if r > 1 {
        panic!("No modular inverse exists for {} mod {}", a, p);
    }
    if t < 0 {
        t += p as i64;
    }
    t as u64
}

impl IntModP {
    pub fn new(i: u64) -> Self {
        let p = get_modulus();
        IntModP { i: i.rem_euclid(p) }
    }

    pub fn copy(&self) -> IntModP {
        IntModP::new(self.i)
    }

    pub fn coerce(&self, value: f64) -> IntModP {
        IntModP::new(value as u64)
    }

    pub fn coerce_to_f64(&self) -> f64 {
        self.i as f64
    }
}

impl IField for IntModP {
    fn a(&self, o: &IntModP) -> IntModP {
        IntModP::new(self.i + o.i)
    }

    fn ae(&mut self, o: &IntModP) {
        let p = get_modulus();
        self.i = (self.i + o.i).rem_euclid(p);
    }

    fn s(&self, o: &IntModP) -> IntModP {
        let p = get_modulus();
        IntModP::new(self.i + p - o.i)
    }

    fn se(&mut self, o: &IntModP) {
        let p = get_modulus();
        self.i = (self.i + p - o.i).rem_euclid(p);
    }

    fn m(&self, o: &IntModP) -> IntModP {
        IntModP::new(self.i * o.i)
    }

    fn me(&mut self, o: &IntModP) {
        let p = get_modulus();
        self.i = (self.i * o.i).rem_euclid(p);
    }

    fn d(&self, o: &IntModP) -> IntModP {
        let p = get_modulus();
        if o.i == 0 {
            panic!("Division by zero in IntModP");
        }
        else {
            let inv = mod_inverse(o.i, p);
            IntModP::new(self.i * inv)
        }
    }
    fn de(&mut self, o: &IntModP) {
        let p = get_modulus();
        if o.i == 0 {
            panic!("Division by zero in IntModP");
        }
        else {
            let inv = mod_inverse(o.i, p);
            self.i = (self.i * inv).rem_euclid(p);
        }
    }

    fn coerce_to_f64(&self) -> f64 {
        self.i as f64
    }

    fn coerce(&self, value: f64) -> IntModP {
        IntModP::new(value as u64)
    }

    fn is_zero(&self) -> bool {
        self.i == 0
    }
    fn is_one(&self) -> bool {
        self.i == 1
    }

    fn zero(&self) -> IntModP {
        IntModP::new(0)
    }
    fn one(&self) -> IntModP {
        IntModP::new(1)
    }
    fn copy(&self) -> IntModP {
        self.copy()
    }
}

impl IOrdered for IntModP {
    fn lt(&self, o: &IntModP) -> bool {
        self.i < o.i
    }

    fn le(&self, o: &IntModP) -> bool {
        self.i <= o.i
    }

    fn gt(&self, o: &IntModP) -> bool {
        self.i > o.i
    }

    fn ge(&self, o: &IntModP) -> bool {
        self.i >= o.i
    }

    fn e(&self, o: &IntModP) -> bool {
        self.i == o.i
    }
}

impl IMath for IntModP {
    fn abs(&self) -> f64 {
        self.i as f64
    }

    fn sqrt(&mut self) {
        panic!("Square root not implemented for IntModP");
    }
}

impl fmt::Display for IntModP {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "IntModP({})", self.i)
    }
}

impl Clone for IntModP {
    fn clone(&self) -> Self {
        self.copy()
    }
}

// Factorize a number
fn factorize(mut n: u64) -> Vec<u64> {
    let mut factors = Vec::new();
    let mut i = 2;
    while i * i <= n {
        if n % i == 0 {
            factors.push(i);
            while n % i == 0 {
                n /= i;
            }
        }
        i += 1;
    }
    if n > 1 {
        factors.push(n);
    }
    factors
}

 fn mod_pow(base: u64, exp: u64, modulus: u64) -> u64 {
        if modulus <= 0 {
            panic!("Modulus must be positive");
        }

        let mut result = 1;
        let mut base = base % modulus;
        let mut exp = exp;

        while exp > 0 {
            if exp % 2 == 1 {
                result = (result * base) % modulus;
            }
            base = (base * base) % modulus;
            exp /= 2;
        }

        result
    }


impl IPrimitiveRoots<IntModP> for IntModP {
    fn primitive_root(&self, n: u64) -> Self {
        let p = get_modulus();
        let factors = factorize(p as u64 - 1);
        for g in 2..p {
            let mut is_root = true;
            for &factor in &factors {
                if mod_pow(g, (p - 1) / factor as u64, p) == 1 {
                    is_root = false;
                    break;
                }
            }
            if is_root {
                return Self::new(g);
            }
        }
        Self::new(0)
    }

    fn pow(&self, exp: u64) -> IntModP {
        let p = get_modulus();
        IntModP::new(mod_pow(self.i, exp as u64, p))
    }

    fn precomputeRootsOfUnity(&self, n: u32, direction: i32) -> Vec<IntModP> {
        let p = get_modulus();
        if (p - 1) % n as u64 != 0 {
            panic!("n must divide p-1 for roots of unity to exist in IntModP");
        }
        let g = self.primitive_root(p);
        let omega = g.pow((p - 1) / (n as u64));
        let mut roots = Vec::with_capacity(n as usize);
        for k in 0..n as i32 {
            let mut exponent: u64 = ((k * direction + (p - 1) as i32) % (p - 1) as i32) as u64;
            if exponent < 0 {
                exponent += (p - 1) as u64;
            }
            roots.push(omega.pow(exponent));
        }
        roots
    }
}

impl Hash for IntModP {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.i.hash(state);
    }
}

impl PartialEq for IntModP {
    fn eq(&self, other: &Self) -> bool {
        self.i == other.i
    }
}

impl Eq for IntModP {}

File: go/generic/int_mod_p.go
package generic

import (
	"fmt"
	"sync"
)

var (
	modulus     uint64
	modulusOnce sync.Once
)

func SetModulus(m uint64) {
	modulusOnce.Do(func() {
		modulus = m
	})
}
func GetModulus() uint64 {
	return modulus
}

// IntModP represents an integer modulo a prime number
type IntModP struct {
	Value uint64
}

// NewIntModP creates a new IntModP instance
func NewIntModP(value uint64) IntModP {
	return IntModP{
		Value: value % modulus,
	}
}

// Modular inverse using Extended Euclidean Algorithm
func modInverse(a, p uint64) uint64 {
	var t, newT int64 = 0, 1
	var r, newR int64 = int64(p), int64(a % p)

	for newR != 0 {
		quotient := r / newR
		t, newT = newT, t-quotient*newT
		r, newR = newR, r-quotient*newR
	}

	if r > 1 {
		panic(fmt.Sprintf("No modular inverse exists for %d mod %d", a, p))
	}
	if t < 0 {
		t += int64(p)
	}
	return uint64(t)
}

// Implement IField interface for IntModP
func (i IntModP) a(o IntModP) IntModP {
	return NewIntModP(i.Value + o.Value)
}

func (i *IntModP) ae(o IntModP) {
	i.Value = (i.Value + o.Value) % modulus
}

func (i IntModP) s(o IntModP) IntModP {
	return NewIntModP(i.Value + modulus - o.Value)
}

func (i *IntModP) se(o IntModP) {
	i.Value = (i.Value + modulus - o.Value) % modulus
}

func (i IntModP) m(o IntModP) IntModP {
	return NewIntModP(i.Value * o.Value)
}

func (i *IntModP) me(o IntModP) {
	i.Value = (i.Value * o.Value) % modulus
}

func (i IntModP) d(o IntModP) IntModP {
	if o.Value == 0 {
		panic("Division by zero in IntModP")
	}
	inv := modInverse(o.Value, modulus)
	return NewIntModP(i.Value * inv)
}

func (i *IntModP) de(o IntModP) {
	if o.Value == 0 {
		panic("Division by zero in IntModP")
	}
	inv := modInverse(o.Value, modulus)
	i.Value = (i.Value * inv) % modulus
}

func (i IntModP) coerceFromInt(v int) IntModP {
	return NewIntModP(uint64(v))
}

func (i IntModP) coerceFromFloat(f float64) IntModP {
	return NewIntModP(uint64(f))
}

func (i IntModP) coerceToFloat() float64 {
	return float64(i.Value)
}

func (i IntModP) isZero() bool {
	return i.Value == 0
}

func (i IntModP) isOne() bool {
	return i.Value == 1
}

func (i IntModP) zero() IntModP {
	return NewIntModP(0)
}

func (i IntModP) one() IntModP {
	return NewIntModP(1)
}

// Implement IMath interface for IntModP
func (i IntModP) abs() IntModP {
	return NewIntModP(i.Value)
}

func (i IntModP) sqrt() IntModP {
	panic("Square root not implemented for IntModP")
}

// Factorize a number into its prime factors
func factorize(n uint64) []uint64 {
	factors := []uint64{}
	for i := uint64(2); i*i <= n; i++ {
		for n%i == 0 {
			factors = append(factors, i)
			n /= i
		}
	}
	if n > 1 {
		factors = append(factors, n)
	}
	return factors
}

// Modular exponentiation
func modPow(base, exp, modulus uint64) uint64 {
	if modulus == 0 {
		panic("Modulus must be positive")
	}

	result := uint64(1)
	base %= modulus

	for exp > 0 {
		if exp%2 == 1 {
			result = (result * base) % modulus
		}
		base = (base * base) % modulus
		exp /= 2
	}

	return result
}

// Implement IPrimitiveRoots interface for IntModP
func (i IntModP) primitiveRoots(n int64) IntModP {
	if n == 0 || n >= int64(modulus) {
		panic("n must be in range [1, p-1]")
	}

	factors := factorize(modulus - 1)
	for g := uint64(2); g < modulus; g++ {
		isRoot := true
		for _, factor := range factors {
			if modPow(g, (modulus-1)/factor, modulus) == 1 {
				isRoot = false
				break
			}
		}
		if isRoot {
			return NewIntModP(g)
		}
	}

	return NewIntModP(0) // No primitive root found
}

func (i IntModP) pow(exp int64) IntModP {
	return NewIntModP(modPow(i.Value, uint64(exp), modulus))
}

func (i IntModP) precomputeRootsOfUnity(n int, direction int) []IntModP {
	if (modulus-1)%uint64(n) != 0 {
		panic("n must divide p-1 for roots of unity to exist in IntModP")
	}

	g := i.primitiveRoots(int64(modulus - 1))
	omega := g.pow(int64((modulus - 1) / uint64(n)))
	roots := make([]IntModP, n)
	for k := 0; k < n; k++ {
		exponent := uint64(k) * uint64(direction) % (modulus - 1)
		roots[k] = omega.pow(int64(exponent))
	}
	return roots
}

// Implement fmt.Stringer interface for IntModP
func (i IntModP) String() string {

	return fmt.Sprintf("IntModP(%d, %d)", i.Value, modulus)
}

// Implement IOrdered interface for IntModP
func (i IntModP) lt(o IntModP) bool {
	return i.Value < o.Value
}

func (i IntModP) le(o IntModP) bool {
	return i.Value <= o.Value
}

func (i IntModP) eq(o IntModP) bool {
	return i.Value == o.Value
}

func (i IntModP) gt(o IntModP) bool {
	return i.Value > o.Value
}

func (i IntModP) ge(o IntModP) bool {
	return i.Value >= o.Value
}

func (i IntModP) copy() IntModP {
	return IntModP{
		Value: i.Value,
	}
}
